### ISSUE TITLE
Cluster IP advertisement is being filtered by route reflector

### ISSUE THREAD

**Issue by swalberg** [projectcalico/confd#199]
I have followed the [guidelines Advertising Kubernetes services over BGP](https://docs.projectcalico.org/master/usage/service-advertisement). My Kubernetes environment has 4 nodes. Two are workers, two are controllers, and the controllers also run route reflectors to which the workers peer along with an external Exabgp router.

We see that the cluster IP is being recognized in the `calico-node` logs:

```
2019-01-04 17:04:35.856 [DEBUG][67] routes.go 246: Advertising local service svc="default/prometheus-server"
```

I see the `/23` route for the cluster IPs in the Exabgp router, but not the `/32` for the service itself. After some debugging, I see the following:

On the worker node that's running a pod, in the `bird_ipam.cfg` file, I see

```
# Export static routes.
if ( net = CLUSTER_NET/23 ) then { accept; }
if ( net = CLUSTER_IP/32 ) then { accept; }
```

on the route reflector, I only see
```
# Export static routes.
if ( net = CLUSTER_NET/23 ) then { accept; }
```

I'm also not sure if it's related to the filtering, but on the exabgp router I only see two entries for the `/23` that come from the route reflector. I would have expected to see them also come from the two worker pods.

Thanks!

**Comment by caseydavenport** [comment]
Yeah, it looks like in a route reflector environment the `calico/node` won't readvertise the `/32` routes it learns for services. 

This is OK for in-cluster nodes since they should be running kube-proxy anyway, but if the RR is peered with something external it needs to know to advertise those /32s. 

I think that either we need to:
- Switch the CLUSTER_NET filter to use `~` so that it includes the /32 routes and accepts them.
- Make RR nodes aware of _all_ local routes in the cluster (i.e., don't filter on whether it matches the node if that node is a RR).

Or, maybe something else.

**Comment by swalberg** [comment]
Option 2 seems like a bit more magic than I'd expect... `~ CLUSTER_NET` is probably the least surprising behavior.

**Comment by caseydavenport** [comment]
Yeah, I agree, and option 1 is probably simpler to implement!

We need to double check that the `~` operator will accept the net itself, or if it will just accept routes within that net. I can't recall of the top of my head what the behavior is.

**Comment by swalberg** [comment]
It seems to... I'm on my RR, and `show ip route` shows all the routes I'd expect. Using the filtering on that:

```
bird> show route filter { if net ~ CLUSTER_NET.0/23 then accept ; }
CLUSTER_NET.192/32  via INTERNAL.7 on ens192 [Node_INTERNAL.7 2019-01-04] * (100/0) [i]
CLUSTER_NET.0/23    blackhole [static1 2019-01-04] * (200)
                   via INTERNAL.6 on ens192 [Node_INTERNAL.6 2019-01-04] (100/0) [i]
                   via INTERNAL.7 on ens192 [Node_INTERNAL.7 2019-01-04] (100/0) [i]
                   via INTERNAL.7 on ens192 [Global_INTERNAL.8 2019-01-04 from INTERNAL.8] (100/0) [i]
                   via INTERNAL.6 on ens192 [Global_INTERNAL.8 2019-01-04 from INTERNAL.8] (100/0) [i]
                   via INTERNAL.8 on ens192 [Global_INTERNAL.8 2019-01-04] (100/0) [i]
bird> show route filter { if net = CLUSTER_NET.0/23 then accept; }
CLUSTER_NET.0/23    blackhole [static1 2019-01-04] * (200)
                   via INTERNAL.6 on ens192 [Node_INTERNAL.6 2019-01-04] (100/0) [i]
                   via INTERNAL.7 on ens192 [Node_INTERNAL.7 2019-01-04] (100/0) [i]
                   via INTERNAL.7 on ens192 [Global_INTERNAL.8 2019-01-04 from INTERNAL.8] (100/0) [i]
                   via INTERNAL.6 on ens192 [Global_INTERNAL.8 2019-01-04 from INTERNAL.8] (100/0) [i]
                   via INTERNAL.8 on ens192 [Global_INTERNAL.8 2019-01-04] (100/0) [i]
                   
```

**Comment by neiljerram** [comment]
Thanks @swalberg @caseydavenport for your report and thoughts on this.

In fact, for the static route part of the export filter, instead of all this complexity:
```
{{- $static_key := "/staticroutes"}}
{{- if ls $static_key}}

  # Export static routes.
  {{- range ls $static_key}}
    {{- $parts := split . "-"}}
    {{- $cidr := join $parts "/"}}
  if ( net = {{$cidr}} ) then { accept; }
  {{- end}}
{{- end}}
```
I think we _should_ only need one line:
```
  if ( net ~ {{$cluster_cidr}} ) then { accept; }
```
But the problem is that the calling code does not generate the `$cluster_cidr` variable, or provide a single etcd key/value pair containing the cluster CIDR.  Instead it generates a list of CIDRs under `/staticroutes`, and we know that the cluster CIDR will always be one of those.

So changing `=` to `~` will work, but the export filter will contain some redundant lines, on nodes with service endpoints.

Note: a key thing to realise about all this is that the templating for the static routes that we actually _try_ to export is elsewhere: https://github.com/projectcalico/confd/blob/master/etc/calico/confd/templates/bird_aggr.cfg.template#L16.

**Comment by caseydavenport** [comment]
@neiljerram yep, that's right. I think the logic that is there currently is overkill for our current use of it, but IIRC it was written with a few other use-cases in mind (which we may or may not implement...) that might end up advertising additional routes that aren't within the clusterIP range and so wouldn't be covered by the one-line solution.

We could certainly revert to the simpler solution for now, but I'm not sure how much it matters that we have the redundant lines.

**Comment by caseydavenport** [comment]
Fix will be in Calico v3.6. This was the PR: https://github.com/projectcalico/confd/pull/206
